File: src/P012/p012.ts
import { countOfDivisors } from "../utils";

const triangleNumbers = function* () {
  for (let t = 0, n = 1; ; n++) {
    t += n;
    yield t;
  }
};

export const solve = (): string => {
  for (const t of triangleNumbers()) {
    if (countOfDivisors(t) > 500) {
      return t.toString();
    }
  }
  throw new Error("No solution found");
};

File: src/P012/p012.go
package problem012

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
)

func Solve() string {
	primes := primeutils.GeneratePrimes(120)
	answer := 0
	for k := 1; ; k++ {
		count1 := primeutils.CountOfDivisors(k, primes)
		count2 := primeutils.CountOfDivisors(2*k-1, primes)
		if count1*count2 > 500 {
			answer = k * (2*k - 1)
			break
		}

		count2 = primeutils.CountOfDivisors(2*k+1, primes)
		if count1*count2 > 500 {
			answer = k * (2*k + 1)
			break
		}
	}
	return strconv.Itoa(answer)
}
